File: JANUARY/11-01-2026/Minimum Window Subsequence.java
class Solution {
    public String minWindow(String s1, String s2) {
        int n = s1.length(), m = s2.length();
        int[][] dp = new int[n + 1][m + 1];

        for (int i = 0; i <= n; i++) dp[i][0] = i;
        for (int j = 1; j <= m; j++) dp[0][j] = -1;

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (s1.charAt(i - 1) == s2.charAt(j - 1))
                    dp[i][j] = dp[i - 1][j - 1];
                else
                    dp[i][j] = dp[i - 1][j];
            }
        }

        int len = Integer.MAX_VALUE, start = -1;
        for (int i = 1; i <= n; i++) {
            if (dp[i][m] != -1) {
                int currLen = i - dp[i][m];
                if (currLen < len) {
                    len = currLen;
                    start = dp[i][m];
                }
            }
        }

        if (start == -1) return "";
        return s1.substring(start, start + len);
    }
}

File: 2026_GFG_POTD/JANUARY/11-01-2026/Minimum Window Subsequence.py
class Solution:
    def minWindow(self, s1, s2):
        n, m = len(s1), len(s2)
        dp = [[-1] * (m + 1) for _ in range(n + 1)]

        for i in range(n + 1):
            dp[i][0] = i

        for i in range(1, n + 1):
            for j in range(1, m + 1):
                if s1[i - 1] == s2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    dp[i][j] = dp[i - 1][j]

        length = float('inf')
        start = -1

        for i in range(1, n + 1):
            if dp[i][m] != -1:
                curr = i - dp[i][m]
                if curr < length:
                    length = curr
                    start = dp[i][m]

        if start == -1:
            return ""
        return s1[start:start + length]